packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp

```diff
@@ -631,79 +685,64 @@ compute ()
-    kh_->destroy_sptrsv_handle();
-#if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA)
-    // CuSparse only supports int type ordinals 
-    // and scalar types of float, double, float complex and double complex
-    if (std::is_same<Kokkos::Cuda, HandleExecSpace>::value &&
-        std::is_same<int, local_ordinal_type>::value &&
-       (std::is_same<scalar_type, float>::value ||
-        std::is_same<scalar_type, double>::value ||
-        std::is_same<scalar_type, Kokkos::complex<float>>::value ||
-        std::is_same<scalar_type, Kokkos::complex<double>>::value))
-    {
-      kh_->create_sptrsv_handle(KokkosSparse::Experimental::SPTRSVAlgorithm::SPTRSV_CUSPARSE, numRows, is_lower_tri);
-    }
-    else
-#endif
-    {
-      kh_->create_sptrsv_handle(KokkosSparse::Experimental::SPTRSVAlgorithm::SEQLVLSCHD_TP1, numRows, is_lower_tri);
-    }
-    KokkosSparse::Experimental::sptrsv_symbolic(kh_.getRawPtr(), ptr, ind, val);
-  }
-  else if (kh_v_nonnull_ && this->isKokkosKernelsSptrsv_) {
-    const bool is_lower_tri = (this->uplo_ == "L") ? true : false;
-
-    for (int i = 0; i < num_streams_; i++) {
-      auto A_crs_i = Teuchos::rcp_dynamic_cast<const crs_matrix_type> (A_crs_v_[i]);
-      auto Alocal_i = A_crs_i->getLocalMatrixDevice();
-      auto ptr_i    = Alocal_i.graph.row_map;
-      auto ind_i    = Alocal_i.graph.entries;
-      auto val_i    = Alocal_i.values;
-   
-      auto numRows_i = Alocal_i.numRows();
-
-      // Destroy existing handle and recreate in case new matrix provided - requires rerunning symbolic analysis
-      kh_v_[i]->destroy_sptrsv_handle();
-#if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA)
-      // CuSparse only supports int type ordinals 
-      // and scalar types of float, double, float complex and double complex
-      if (std::is_same<Kokkos::Cuda, HandleExecSpace>::value &&
-          std::is_same<int, local_ordinal_type>::value &&
-         (std::is_same<scalar_type, float>::value ||
-          std::is_same<scalar_type, double>::value ||
-          std::is_same<scalar_type, Kokkos::complex<float>>::value ||
-          std::is_same<scalar_type, Kokkos::complex<double>>::value))
-      {
-        kh_v_[i]->create_sptrsv_handle(KokkosSparse::Experimental::SPTRSVAlgorithm::SPTRSV_CUSPARSE, numRows_i, is_lower_tri);
-      }
-      else
-#endif
-      {
-        kh_v_[i]->create_sptrsv_handle(KokkosSparse::Experimental::SPTRSVAlgorithm::SEQLVLSCHD_TP1, numRows_i, is_lower_tri);
+// NOTE (Nov-09-2022):
+// For Cuda >= 11.3 (using cusparseSpSV), always call symbolic during compute
+// even when matrix values are changed with the same sparsity pattern.
+// For Cuda >= 12.1 has a new cusparseSpSV_updateMatrix function just for updating the
+// values that is substantially faster.
+// This would all be much better handled via a KokkosSparse::Experimental::sptrsv_numeric(...)
```

Comment:
@vbrunini Just to make sure. You mean sptrsv_symbolic(...), right?

Reply:
No, I mean numeric. I think there need to be 3 phases to the Kokkos Kernels sptrsv interface rather than the current 2, to match the Ifpack2 preconditioner phases:
1) sptrsv_symbolic aligns with initialize() and only needs to be run if the graph changes, it should not depend on the actual matrix values (though awkwardly the cusparse interface does want them to exist). For (current) cusparse this maps to cusparse_SpSV_buffer_size and cusparse_SpSV_analysis.
2) sptrsv_numeric (which doesn't currently exist) aligns with compute() and needs to be called when the matrix values (but not graph) change. For (current) cusparse this maps to cusparse_SpSV_updateMatrix().
3) sptrsv_solve() which maps to apply().